test/rpc_test.py

```diff
@@ -1076,6 +1076,50 @@ def test_call_method_on_rref(self):
 
         self.assertEqual(result, sum(vals))
 
+    @dist_init(setup_rpc=False)
+    @requires_process_group_agent("PROCESS_GROUP rpc backend specific test, skip")
+    def test_rpc_agent_destructor(self):
+        # This test ensures that the rpc agent destructor safely shuts down the
+        # local RPC framework even if rpc.shutdown is not called
+        rpc.init_rpc(
+            self_name="worker%d" % self.rank,
+            backend=rpc.backend_registry.BackendType[TEST_CONFIG.rpc_backend_name],
+            self_rank=self.rank,
+            worker_name_to_id=self.worker_name_to_id,
+            init_method=self.init_method,
+        )
+
+    @dist_init(setup_rpc=False)
+    @requires_process_group_agent("PROCESS_GROUP rpc backend specific test, skip")
+    def test_rpc_local_shutdown(self):
+        rpc.init_rpc(
+            self_name="worker%d" % self.rank,
+            backend=rpc.backend_registry.BackendType[TEST_CONFIG.rpc_backend_name],
+            self_rank=self.rank,
+            worker_name_to_id=self.worker_name_to_id,
+            init_method=self.init_method,
+        )
+        rpc.shutdown()
+
+    @dist_init(setup_rpc=False)
+    @requires_process_group_agent("PROCESS_GROUP rpc backend specific test, skip")
+    def test_rpc_join_and_shutdown(self):
+        # This tests ensures that both rpc.join_rpc() and rpc.shutdown() can be called without errors being raised due to attempting to shut down multiple times.
+        rpc.init_rpc(
+            self_name="worker%d" % self.rank,
+            backend=rpc.backend_registry.BackendType[TEST_CONFIG.rpc_backend_name],
+            self_rank=self.rank,
+            worker_name_to_id=self.worker_name_to_id,
+            init_method=self.init_method,
+        )
+        rpc.join_rpc()
+        rpc.shutdown()
```

Comment:
Question, in this latest version, `join_rpc()` not longer shutdown the agent? And will it block until everyone the rendezvous calls rpc.join_rpc()?

Reply:
`join_rpc` will still shutdown everything, the key difference is that it'll also send a message to other workers telling them to shutdown, `localShutdown` won't. Existing workflows shouldn't be affected by this change.

> And will it block until everyone the rendezvous calls rpc.join_rpc()?

If you're referring to whether `rpc.join_rpc` will block until all workers call it, yes, this is true (and not changed by this PR)

Reply:
Probably #30050, just because the CI will likely finish first and I'd like to not block that PR any longer (API name changes are especially prone to breakages/merge conflicts the longer they are out there)